In [ ]:
!pip install pyodbc
!apt-get update
!apt-get install -y unixodbc unixodbc-dev
# download Microsoft SQL Server ODBC driver 17
!curl https://packages.microsoft.com/keys/microsoft.asc | apt-key add -
!curl https://packages.microsoft.com/config/ubuntu/$(lsb_release -rs)/prod.list | tee /etc/apt/sources.list.d/mssql-release.list
!apt-get update
!ACCEPT_EULA=Y apt-get install -y msodbcsql17

import requests
# GET IP
ip_address = requests.get('https://api64.ipify.org').text
print("Your IP address is:", ip_address)

Hit:1 http://archive.ubuntu.com/ubuntu jammy InRelease
Hit:2 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease
Get:3 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]
Hit:4 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease
Hit:5 https://packages.microsoft.com/ubuntu/22.04/prod jammy InRelease
Get:6 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Hit:7 http://archive.ubuntu.com/ubuntu jammy-backports InRelease
Hit:8 https://r2u.stat.illinois.edu/ubuntu jammy InRelease
Hit:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Hit:10 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Hit:11 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Fetched 257 kB in 2s (170 kB/s)
Reading package lists... Done
W: https://packages.microsoft.com/ubuntu/22.04/prod/dists/jammy/InRelease: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.

In [ ]:
from sqlalchemy import create_engine
import sqlalchemy
!pip install pyodbc
!pip install sqlalchemy

# Define connection details
username = 'sqladmin'  # Replace with your Azure SQL admin username
password = 'MINIproject11'        # Replace with your Azure SQL password
server = 'zienzhu2-server.database.windows.net'
database = 'mini-project-db'

# Connection string for Azure SQL Server
connection_string = f"mssql+pyodbc://{username}:{password}@{server}:1433/{database}?driver=ODBC+Driver+17+for+SQL+Server"
engine = create_engine(connection_string)

# Load the ipython-sql extension
%load_ext sql

# Connect to the database
%sql $connection_string

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [ ]:
# Checking if all tables got imported in the database
%%sql
SELECT TABLE_NAME
FROM INFORMATION_SCHEMA.TABLES
WHERE TABLE_TYPE = 'BASE TABLE';

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


KeyError: 'DEFAULT'

In [ ]:
# Checking for the sample data from each table

In [ ]:
%%sql
SELECT TOP 2* FROM business_df

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[(0, 'Pns2l4eNsfO8kk83dixA6A', 'Abby Rappoport, LAC, CMQ', '1616 Chapala St, Ste 2', 'Santa Barbara', 'CA', '93101', 34.4266787, -119.7111968, 5.0, 7, 0, '{"ByAppointmentOnly": "True"}', 'Doctors, Traditional Chinese Medicine, Naturopathic/Holistic, Acupuncture, Health & Medical, Nutritionists', None),
 (1, 'mpf3x-BjTdTEA3yCZrAYPw', 'The UPS Store', '87 Grasso Plaza Shopping Center', 'Affton', 'MO', '63123', 38.551126, -90.335695, 3.0, 15, 1, '{"BusinessAcceptsCreditCards": "True"}', 'Shipping Centers, Local Services, Notaries, Mailbox Centers, Printing Services', '{"Monday": "0:0-0:0", "Tuesday": "8:0-18:30", "Wednesday": "8:0-18:30", "Thursday": "8:0-18:30", "Friday": "8:0-18:30", "Saturday": "8:0-14:0"}')]

In [ ]:
%%sql
SELECT TOP 2* FROM review_df

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[('---4VcQZzy_vIIifUDqxsg', 'EopuF3BhVXAGJWEje_TJ-g', 'X-dPdw1PluJnT-fpiWrfKw', 1.0, 1, 0, 0, "After being seated for over 15 minutes without seeing our waitress, we should have known it was not going to be a good experience.  I will start by s ... (486 characters truncated) ... gotten worse every time. I genuinely wanted this establishment to be successful, as I said the food is good, but the rest of the meal was a disaster.", '2018-09-16 00:38:13'),
 ('---aZVSWnQtOP_12QzIt8g', 'Zr2x9lE0thdxVXvqBnl-NA', 'SCjUeZfjafJ88fWWQjRh2w', 5.0, 0, 0, 0, "We live multiple hours away, but we come to philly often. We love coming here when we're in the mood for Mexican. Consistently delicious. Our last se ... (276 characters truncated) ... 've ordered in between trips have always been consistent. I have no plans of finding anywhere else to enjoy tequila and Mexican food while in philly.", '2018-06-16 02:49:40')]

In [ ]:
%%sql
SELECT TOP 1* FROM user_df

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[(397935, '---1lKK3aKOuomHnwAkAow', 'Monera', 266, '2007-06-04 01:37:45', 510, 181, 206, '2010,2011,2012', 'qewG3X2O4X6JKskxyyqFwQ, ci7Q6NvXQ8UZrBHusVX18g, bW2AQqcSPascYvKXsF0S-g, H3wPBzQCKCKiHcV7GZK4kg, RiU99Ze8UEoUutHMl73HIg, 3vEdsqwOPevuoNvHphkQyQ, N7Jzf ... (20988 characters truncated) ... 338rg, JL0SWJAcngxhoRSMgKWF_w, QgL-vJ74cBMqllv1HFecsQ, z4KgHjl1ryNgT_h1s5um5w, aPcF4QD3X6NOrR-_SMSsug, Ie2bZZdv02AAgx_niX8g9A, z6yd1BoLPTAEUPPYDKHazA', 17, 3.93, 2, 3, 2, 1, 0, 5, 9, 9, 9, 9, 0)]

In [ ]:
%%sql
SELECT TOP 2* FROM checkin_df

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[(0, '---kPU91CF4Lq2-WlRu9Lw', '2020-03-13 21:10:56, 2020-06-02 22:18:06, 2020-07-24 22:42:27, 2020-10-24 21:36:13, 2020-12-09 21:23:33, 2021-01-20 17:34:57, 2021-04-30 21:02:03, 2021-05-25 21:16:54, 2021-08-06 21:08:08, 2021-10-02 15:15:42, 2021-11-11 16:23:50'),
 (1, '--0iUa4sNDFiZFrAdIWhZQ', '2010-09-13 21:43:09, 2011-05-04 23:08:15, 2011-07-18 22:30:31, 2012-09-07 20:28:50, 2013-03-27 15:57:36, 2013-08-13 00:31:34, 2013-08-13 00:31:48, 2013-09-23 17:39:38, 2013-11-18 06:34:08, 2014-04-12 23:04:47')]

In [ ]:
%%sql
SELECT TOP 2* FROM tip_df

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[('---r61b7EpVPkb4UVme5tA', '9PZxjhTIU7OgPIzuGi89Ew', 'Stopped in for lunch. Nice spot! Will definitely be back.', '2016-01-17 00:08:25', 0),
 ('--_H9j6ggxvqhh9nPofZwg', '0-ViWW1_-_TKIf-o8o8bPw', '1/2 price bottles of wine Wednesdays', '2016-07-22 17:23:08', 0)]

In [ ]:
# Writing Queries

# Query 1 & 2 - Identifying the right city, to open the restaurant business

Query 1- Pulling out cities with maximum number of restaurants


In [ ]:
%%sql
SELECT TOP 20 city, count(DISTINCT business_id) as Restaurant_count
FROM business_df
WHERE UPPER(business_df.categories) LIKE ('%RESTAURANT%')
GROUP BY city
ORDER BY Restaurant_count DESC;

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Philadelphia', 5857),
 ('Tampa', 2968),
 ('Indianapolis', 2863),
 ('Nashville', 2506),
 ('Tucson', 2473),
 ('New Orleans', 2262),
 ('Edmonton', 2168),
 ('Saint Louis', 1791),
 ('Reno', 1290),
 ('Boise', 850),
 ('Santa Barbara', 767),
 ('Clearwater', 679),
 ('Wilmington', 622),
 ('St. Louis', 544),
 ('Metairie', 522),
 ('Saint Petersburg', 491),
 ('Franklin', 444),
 ('St. Petersburg', 405),
 ('Sparks', 334),
 ('Brandon', 327)]

In [ ]:
Query 2- Checking the average ratings in each city

In [ ]:
%%sql
SELECT TOP 20 city, count(review_df.stars) as reviewer_count,round(AVG(review_df.stars),2) AS average_rating
FROM review_df
JOIN business_df ON review_df.business_id = business_df.business_id
WHERE UPPER(business_df.categories) LIKE ('%RESTAURANT%')
AND city IN ('Philadelphia','Tampa','Indianapolis','Nashville','Tucson','New Orleans','Edmonton','Saint Louis','Reno','Boise',
'Santa Barbara','Clearwater','Wilmington','St. Louis','Metairie','Saint Petersburg','Franklin','St. Petersburg','Sparks','Brandon')
GROUP BY city
ORDER BY average_rating DESC;

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Santa Barbara', 162283, 3.98),
 ('New Orleans', 476572, 3.95),
 ('Saint Petersburg', 52885, 3.93),
 ('Clearwater', 56443, 3.89),
 ('St. Petersburg', 36179, 3.88),
 ('Nashville', 325787, 3.87),
 ('Indianapolis', 250642, 3.86),
 ('St. Louis', 46010, 3.86),
 ('Boise', 66644, 3.85),
 ('Tampa', 303698, 3.84),
 ('Saint Louis', 178460, 3.84),
 ('Philadelphia', 687571, 3.81),
 ('Reno', 200672, 3.78),
 ('Tucson', 249863, 3.74),
 ('Brandon', 30145, 3.7),
 ('Franklin', 38136, 3.7),
 ('Edmonton', 64042, 3.66),
 ('Wilmington', 31699, 3.65),
 ('Metairie', 42186, 3.64),
 ('Sparks', 40186, 3.59)]

*Tucson has a good number of restaurant i.e. 2473, but the average rating is quite low, i.e. 3.74, though there are ~250K reviews- which indicates that Tucson has a room of growth.*

# Query 3 -Top Restaurant Categories by Average Rating

In [ ]:
%%sql
SELECT TOP 20 categories, Round(AVG(review_df.stars),2) AS average_rating
FROM review_df
JOIN business_df ON review_df.business_id = business_df.business_id
WHERE UPPER(business_df.categories) LIKE ('%RESTAURANT%') AND city = 'Tucson'
GROUP BY categories
ORDER BY average_rating DESC;

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Brasseries, Comfort Food, Restaurants, Food, Pizza, Sandwiches, Delis', 5.0),
 ('Mexican, Cafes, Restaurants, Breakfast & Brunch', 5.0),
 ('Food Trucks, Mexican, Restaurants, Hot Dogs, Food', 5.0),
 ('Caterers, Restaurants, Cultural Center, Cafes, Arts & Entertainment, Event Planning & Services, Local Services, Venues & Event Spaces', 5.0),
 ('Event Planning & Services, Caterers, Food, Sandwiches, American (Traditional), Restaurants, Food Trucks, Barbeque, Street Vendors', 5.0),
 ('Street Vendors, Food Trucks, Food, Mexican, Restaurants', 5.0),
 ('Burgers, Food Trucks, Restaurants, Food', 5.0),
 ('Shopping, Food, Wholesale Stores, Restaurants, Specialty Food, Cafes, Coffee Roasteries, Coffee & Tea', 5.0),
 ('Tattoo, Cafes, Beauty & Spas, Restaurants', 5.0),
 ('Puerto Rican, Caribbean, Restaurants, Desserts, Food Trucks, Food', 5.0),
 ('Food, Soup, Desserts, Gluten-Free, Vegan, Bakeries, Sandwiches, Restaurants', 5.0),
 ('American (Traditional), Food, Food Trucks, Restaurants, Barbeque', 5.0),
 ('Shaved Ice, Food Stands, Restaurants, Street Vendors, Food Trucks, Food, Desserts', 5.0),
 ('Trainers, Active Life, Gyms, Health Markets, Restaurants, Food Delivery Services, Specialty Food, Food, Fitness & Instruction, Food Trucks, Event Planning & Services, Juice Bars & Smoothies, Personal Chefs', 5.0),
 ('Restaurants, Cafes, Juice Bars & Smoothies, Food, Street Vendors, Coffee & Tea', 4.97),
 ('Beauty & Spas, Skin Care, Hair Removal, Massage, Waxing, Makeup Artists, Restaurants, Reflexology, Health & Medical, Thai, Massage Therapy, Day Spas', 4.96),
 ('Restaurants, Vegan', 4.92),
 ('Restaurants, Mexican, Food Trucks, Food', 4.92),
 ('Cafes, American (Traditional), Breakfast & Brunch, Restaurants', 4.92),
 ('Tex-Mex, Restaurants, Seafood, Mexican', 4.92)]

# Query 4 - Checking which restaurant category is more reviewed

In [ ]:
%%sql
SELECT TOP 10 categories, COUNT(*) AS review_count
FROM review_df
JOIN business_df ON review_df.business_id = business_df.business_id
WHERE UPPER(business_df.categories) LIKE ('%RESTAURANT%') AND city = 'Tucson'
GROUP BY categories
ORDER BY review_count DESC;

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Restaurants, Mexican', 6258),
 ('Mexican, Restaurants', 5913),
 ('Restaurants, Cocktail Bars, Bars, Comfort Food, American (New), Sandwiches, Nightlife, Breakfast & Brunch', 2230),
 ('Restaurants, American (New)', 2078),
 ('Tapas/Small Plates, Desserts, Cocktail Bars, Nightlife, Bars, Tacos, Mexican, Food, Restaurants, Salad', 1617),
 ('Chinese, Restaurants', 1349),
 ('Restaurants, Chinese', 1340),
 ('American (Traditional), Restaurants, Mexican, Latin American', 1331),
 ('Ice Cream & Frozen Yogurt, American (New), Nightlife, Sandwiches, Bars, Restaurants, Salad, American (Traditional), Desserts, Food', 1329),
 ('Italian, Restaurants', 1327)]

*Mexican restaurants are highly reviewed and the average rating is positive, which suggests that people in that region prefers Mexican food. Hence, I would suggest opening a Mexican retaurant in Tucson.*

# Query 5 & 6 - Checking the common words in positive reviews (ratings >=4) and negative reviews (ratings <= 2) Rating 3 is assumed to be a mediocre restaurant

In [ ]:
%%sql

SELECT TOP 20 word, COUNT(*) AS frequency
FROM (
    SELECT value AS word, review_df.business_id
    FROM review_df
    CROSS APPLY STRING_SPLIT(text, ' ')  -- This splits the 'text' column into words
    WHERE stars >= 4
) AS positive_words
JOIN business_df ON positive_words.business_id = business_df.business_id
WHERE UPPER(business_df.categories) LIKE '%RESTAURANT%'
AND city = 'Tucson'
AND UPPER(business_df.categories) LIKE '%MEXICAN%'
AND upper(word) NOT IN (' ','I','THE','AND','A','TO','IS','WAS','OF','IN','BUT','ON','FOR','IT','WITH','THIS','THAT','WE','ME','MINE',
'YOU','THEY','ARE','HAD','MY','IF','AN','HAVE','WERE','SO','GOOD','GREAT','VERY','NOT','AS','BE','AT','THEIR','ALL',
'OUR','HERE','LIKE','BEST','ONE','JUST','FROM','WILL','GET','THERE','REALLY','OUT','GO','LOVE','ALWAYS','ALSO','SOME','WHICH','BEEN','WHEN','TIME'
,'YOUR','WOULD','SHOULD','MIGHT','HATE','POOR','OR','HAS','ABOUT','CAN','BACK','NO','SHE','HE','US','GOT','EVEN','NEVER','AFTER','ONLY','BECAUSE','CAME','UP','DOWN','WHAT',
'BY','THEN','SAID','WENT','OTHER','IT_S','%DON''T%','DO NOT','%DID''NT%','ASKED','MORE','ASKED','TOLD','MENTIONED','%I''VE%','TOOK','DID','THEM','HER','COULD','OVER','%IT''S%',
'-','TRY','LITTLE')
GROUP BY word
ORDER BY frequency DESC;

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[('food', 20086),
 ('place', 13818),
 ('Mexican', 9544),
 ("it's", 7796),
 ('salsa', 7760),
 ('service', 7607),
 ('tacos', 7133),
 ("I've", 5256),
 ('definitely', 5222),
 ('restaurant', 5093),
 ('Tucson', 4926),
 ('ordered', 4867),
 ('delicious', 4469),
 ('fresh', 4351),
 ('Nice', 4304),
 ('carne', 4273),
 ('taco', 4244),
 ('good.', 4219),
 ('made', 4156),
 ('menu', 4118)]

*The common words would allow us to know what people like the most, or appreciate the most in positive reviews. To be implemented while setting up the business - Common words - Food, Place, Salsa, Service, Menu*

In [ ]:
%%sql

SELECT TOP 20 word, COUNT(*) AS frequency
FROM (
    SELECT value AS word, review_df.business_id
    FROM review_df
    CROSS APPLY STRING_SPLIT(text, ' ')  -- This splits the 'text' column into words
    WHERE stars < 3
) AS negative_words
JOIN business_df ON negative_words.business_id = business_df.business_id
WHERE UPPER(business_df.categories) LIKE '%RESTAURANT%'
AND city = 'Tucson'
AND UPPER(business_df.categories) LIKE '%MEXICAN%'
AND upper(word) NOT IN (' ','I','THE','AND','A','TO','IS','WAS','OF','IN','BUT','ON','FOR','IT','WITH','THIS','THAT','WE','ME','MINE',
'YOU','THEY','ARE','HAD','MY','IF','AN','HAVE','WERE','SO','GOOD','GREAT','VERY','NOT','AS','BE','AT','THEIR','ALL',
'OUR','HERE','LIKE','BEST','ONE','JUST','FROM','WILL','GET','THERE','REALLY','OUT','GO','LOVE','ALWAYS','ALSO','SOME','WHICH','BEEN','WHEN','TIME'
,'YOUR','WOULD','SHOULD','MIGHT','HATE','POOR','OR','HAS','ABOUT','CAN','BACK','NO','SHE','HE','US','GOT','EVEN','NEVER','AFTER','ONLY','BECAUSE','CAME','UP','DOWN','WHAT',
'BY','THEN','SAID','WENT','OTHER','IT_S','%DON''T%','DO NOT','%DID''NT%','ASKED','MORE','ASKED','TOLD','MENTIONED','%I''VE%','TOOK','DID','THEM','HER','COULD','OVER','%IT''S%')
GROUP BY word
ORDER BY frequency DESC;

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[('food', 9339),
 ('place', 4510),
 ('ordered', 3700),
 ('service', 3680),
 ('order', 3121),
 ('Mexican', 2765),
 ("Don't", 2642),
 ("didn't", 2537),
 ('restaurant', 2354),
 ("It's", 2176),
 ('minutes', 2137),
 ('salsa', 2046),
 ('tacos', 1810),
 ('two', 1784),
 ('do', 1783),
 ('taco', 1755),
 ('going', 1746),
 ('food.', 1709),
 ("I'm", 1687),
 ('it.', 1681)]

*The common words would allow us to know what people dislike the most in negative reviews. To be avoided while setting up the business. Based on our findings, it looks like people have complained about the service, order, food and place, so while building we should not compromise on this factors*

# Query 7 - Identifying the days on which Restaurant should be open




In [ ]:
%%sql
SELECT
    CASE
        WHEN DATEPART(WEEKDAY, CONVERT(DATETIME,
            SUBSTRING(checkin_df.date, 1, CHARINDEX(',', checkin_df.date + ',') - 1))) = 1 THEN 'Sunday'
        WHEN DATEPART(WEEKDAY, CONVERT(DATETIME,
            SUBSTRING(checkin_df.date, 1, CHARINDEX(',', checkin_df.date + ',') - 1))) = 2 THEN 'Monday'
        WHEN DATEPART(WEEKDAY, CONVERT(DATETIME,
            SUBSTRING(checkin_df.date, 1, CHARINDEX(',', checkin_df.date + ',') - 1))) = 3 THEN 'Tuesday'
        WHEN DATEPART(WEEKDAY, CONVERT(DATETIME,
            SUBSTRING(checkin_df.date, 1, CHARINDEX(',', checkin_df.date + ',') - 1))) = 4 THEN 'Wednesday'
        WHEN DATEPART(WEEKDAY, CONVERT(DATETIME,
            SUBSTRING(checkin_df.date, 1, CHARINDEX(',', checkin_df.date + ',') - 1))) = 5 THEN 'Thursday'
        WHEN DATEPART(WEEKDAY, CONVERT(DATETIME,
            SUBSTRING(checkin_df.date, 1, CHARINDEX(',', checkin_df.date + ',') - 1))) = 6 THEN 'Friday'
        WHEN DATEPART(WEEKDAY, CONVERT(DATETIME,
            SUBSTRING(checkin_df.date, 1, CHARINDEX(',', checkin_df.date + ',') - 1))) = 7 THEN 'Saturday'
    END AS weekday_of_week,
    COUNT(*) AS checkin_count
FROM
    checkin_df
JOIN
    business_df ON checkin_df.business_id = business_df.business_id
WHERE
    business_df.categories LIKE '%RESTAURANT%'
    AND city = 'Tucson'
    AND business_df.categories LIKE '%MEXICAN%'
GROUP BY
    DATEPART(WEEKDAY, CONVERT(DATETIME,
        SUBSTRING(checkin_df.date, 1, CHARINDEX(',', checkin_df.date + ',') - 1)))
ORDER BY
    checkin_count DESC;

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Saturday', 117),
 ('Friday', 84),
 ('Sunday', 76),
 ('Tuesday', 64),
 ('Wednesday', 60),
 ('Thursday', 57),
 ('Monday', 48)]

*Insight - If we need to open restaurant in 6/7 days a week, we can keep Mondays closed as it has minimum checkin.*

# Query 8 - Identifying the hours during which Restaurant should be open

In [ ]:
%%sql
SELECT
    DATEPART(HOUR, CONVERT(DATETIME,
        SUBSTRING(checkin_df.date, 1, CHARINDEX(',', checkin_df.date + ',') - 1))) AS hour_of_day,
    COUNT(*) AS checkin_count
FROM
    checkin_df
JOIN
    business_df ON checkin_df.business_id = business_df.business_id
WHERE
    business_df.categories LIKE '%RESTAURANT%'
    AND city = 'Tucson'
    AND business_df.categories LIKE '%MEXICAN%'
GROUP BY
    DATEPART(HOUR, CONVERT(DATETIME,
        SUBSTRING(checkin_df.date, 1, CHARINDEX(',', checkin_df.date + ',') - 1)))
ORDER BY
    checkin_count DESC;

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[(19, 58),
 (1, 58),
 (2, 57),
 (18, 50),
 (20, 36),
 (23, 32),
 (3, 31),
 (0, 30),
 (22, 28),
 (17, 23),
 (21, 22),
 (4, 17),
 (5, 13),
 (6, 11),
 (15, 10),
 (16, 9),
 (7, 6),
 (14, 5),
 (9, 4),
 (12, 2),
 (8, 2),
 (13, 1),
 (10, 1)]

*Insight- The time when the restaurant should be open are late evenings like 5:00 pm till 3:00 am. As other hours, the Mexcian restaurants aren't seeing a lot of checkin.*

# Query 9 - Identifying the Most Engaged Users (Top Reviewers) in Mexican Restaurant Category in Tucson

In [ ]:
%%sql
SELECT TOP 10 user_id, COUNT(*) AS total_reviews, AVG(review_df.stars) AS avg_user_rating
FROM review_df
JOIN business_df ON review_df.business_id = business_df.business_id
WHERE UPPER(business_df.categories) LIKE ('%RESTAURANT%')
AND UPPER(business_df.categories) LIKE ('%MEXICAN%')
AND city = 'Tucson'
GROUP BY user_id
ORDER BY total_reviews DESC, avg_user_rating DESC;

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[('EAHCCgMFoJxlOPcCPQ91SQ', 81, 3.802469135802469),
 ('BrH27yHhgDEylDZfFdJvSQ', 78, 2.717948717948718),
 ('-rGU3wo4fgEnGRboYdMTjw', 67, 3.582089552238806),
 ('2YKkLFeOx-0zRcWp0KUv_Q', 67, 3.3582089552238807),
 ('PAc93PtEbYDtytBQ9Dyjug', 66, 3.757575757575758),
 ('dvu83QWKsnnbGvNVQ3mv7A', 66, 3.742424242424242),
 ('6ObFF8-uKnOAlXuSH4TlyQ', 65, 4.1692307692307695),
 ('xWmYN57XXZbg0LOK8WbbFQ', 63, 4.825396825396825),
 ('leMOR7VSm5z-0r60iJ90EA', 61, 3.8360655737704916),
 ('9Y1YkloHk2MAE3hkwYdFKA', 58, 3.603448275862069)]

*Insight - We can identify the top users, and before opening the restaurant take their feedback on the food, the ambience and service, which would help us in improving the overall service, before inaugration, and these users can help the restaurant business as an influencer or through their Social Media Posts.*

# Query 10 - Checking customer engagement, as in do they tip or not

In [ ]:
%%sql
SELECT business_df.city,
Round(AVG(user_df.average_stars),2) AS avg_user_rating, SUM(tip_df.compliment_count) AS total_compliments, COUNT(DISTINCT tip_df.business_id) AS businesses_tipped
FROM user_df
JOIN tip_df ON user_df.user_id = tip_df.user_id
JOIN business_df ON tip_df.business_id = business_df.business_id
WHERE UPPER(business_df.categories) LIKE '%RESTAURANT%' AND UPPER(business_df.categories) LIKE '%MEXICAN%'
AND business_df.city = 'Tucson'
GROUP BY business_df.city
ORDER BY avg_user_rating DESC;

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Tucson', 3.8, 118, 471)]

*Insight -  The customers in Tucson city are enthusiastic and engaging with restaurant, they do provide tip/compliments*

# *As a result, we would suggest opening a Mexican cuisine restaurant in Tucson city, which opens for 6 days a week (Tuesday -Sunday), in late evening hours, i.e. 5:00 pm - 3:00 a.m., and the top users can be asked for feedback for a good restaurant.*

# Query 11 - Apart from Tucson, other cities, with very few restaurants can be targeted, if the population of that region is good, this would give a first-mover advantage and would result in a profitable business

In [ ]:
%%sql
SELECT city, state, AVG(stars) AS avg_rating, COUNT(business_id) AS Total_Restaurants
FROM business_df
WHERE categories LIKE '%Restaurant%'
GROUP BY city, state
HAVING AVG(stars) > 4 AND COUNT(business_id) < 5;

 * mssql+pyodbc://sqladmin:***@zienzhu2-server.database.windows.net:1433/mini-project-db?driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Tinicum', 'PA', 4.5, 1),
 ('Picture Rocks', 'AZ', 5.0, 1),
 ('Saintt Petersburg', 'FL', 4.5, 1),
 ('Bridgeton', 'NJ', 4.5, 1),
 ('Birchrunville', 'PA', 4.5, 2),
 ('Glendale', 'MO', 5.0, 1),
 ('Montgomery', 'PA', 4.5, 1),
 ('Erial', 'NJ', 4.5, 1),
 ('Downtown', 'LA', 4.5, 1),
 ('Mont Clare', 'PA', 4.5, 2),
 ('Twn N Cntry', 'FL', 4.5, 1),
 ('Aston', 'DE', 4.5, 1),
 ('West Chester PA', 'PA', 4.5, 1),
 ('St.Ann', 'MO', 4.5, 1),
 ('Medford Lakes', 'NJ', 4.25, 4),
 ('Bucktown', 'LA', 4.5, 1),
 ('Trainer', 'PA', 4.5, 1),
 ('Mount Royal', 'NJ', 4.5, 1),
 ('LOWER PROVIDENCE', 'PA', 5.0, 1),
 ('Wanamaker', 'IN', 4.5, 1),
 ('Pinecrest West Park', 'FL', 4.5, 1),
 ('Blooming Glen', 'PA', 4.5, 1),
 ('Westtown', 'PA', 4.5, 2),
 ('East St Louis', 'IL', 4.25, 2),
 ('Hilltown', 'PA', 5.0, 1),
 ('Sassamansville', 'PA', 5.0, 1),
 ('St. Peters', 'MO', 4.5, 2),
 ('St. Louis', 'IL', 5.0, 1),
 ('Masaryktown', 'FL', 4.5, 1),
 ('Rosewood Heights', 'IL', 5.0, 1),
 ('Pass-a-Grille Beach', 'FL', 4.5, 1),
 ('Wallingford', 'PA', 5.0, 1),
 ('Belmont Hills', 'PA', 4.5, 1),
 ('Roebling', 'NJ', 4.125, 4),
 ('Ozona', 'FL', 4.333333333333333, 3),
 ('Freehold', 'NJ', 4.5, 1),
 ('Madison', 'IL', 4.5, 1),
 ('Dade City', 'FL', 4.166666666666667, 3),
 ('Holicong', 'PA', 4.5, 2),
 ('Wilmington Manor', 'DE', 4.5, 2),
 ('Lawrence Township', 'IN', 4.5, 1),
 ('St.  Charles', 'MO', 5.0, 1),
 ('Waterford Works', 'NJ', 4.5, 2),
 ('Hilltop', 'NJ', 4.5, 1),
 ('Saint Bernard', 'LA', 4.75, 2),
 ('Kimberton', 'PA', 4.333333333333333, 3),
 ("Town 'n' Country", 'FL', 4.625, 4),
 ('North Coventry Township', 'PA', 4.5, 1),
 ('Greater Northdale', 'FL', 4.5, 1),
 ('Balm', 'FL', 4.5, 1),
 ('Blvd', 'MO', 4.5, 1),
 ('Newfield', 'NJ', 4.25, 4),
 ('Caln', 'PA', 4.5, 1),
 ('Sumneytown', 'PA', 4.166666666666667, 3),
 ('Nampa', 'ID', 4.5, 1),
 ('Drexel Heights', 'AZ', 4.5, 1),
 ('Bryn Athyn', 'PA', 5.0, 1),
 ('Kalispell', 'MT', 5.0, 1),
 ('Cedars', 'PA', 4.125, 4),
 ('Truckee', 'CA', 5.0, 1),
 ('Glenoldan', 'PA', 4.5, 1),
 ('Fairmont City', 'IL', 4.25, 2),
 ('Winchester', 'MO', 4.25, 2),
 ('East Falls', 'PA', 4.5, 1),
 ('Pennsville Township', 'NJ', 4.166666666666667, 3),
 ('Lutz fl', 'FL', 5.0, 1),
 ('Lester', 'PA', 4.333333333333333, 3),
 ('Town n Country', 'FL', 4.5, 1),
 ('Tierre Verde', 'FL', 4.5, 1),
 ('Berry Hill', 'TN', 4.333333333333333, 3),
 ('Gwynedd Valley', 'PA', 4.5, 1),
 ('Chichester', 'PA', 4.25, 2),
 ('Gladwyne', 'PA', 4.166666666666667, 3)]